### DISCUSSION TITLE
How to hide a custom resource?

### DISCUSSION THREAD

**Discussion by Kralizek** [microsoft/aspire#5704] [Q&A]
I am working on a custom resource that hosts an emulator of a piece of infrastructure that other teams will be using (a custom OIDC server based on Identity Server).

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var identityManager = builder.AddIdentityManagerContainer()
    .AddOpenIdIdentityResource()
    .AddProfileIdentityResource()
    .AddScope("custom", emphasize: true, required: true)
    .AddApiResource("custom", "Custom", scopes: ["custom"], claims: ["sub", "my_custom_claim"])
    .AddUser("alice", "alice")
    .AddUser("bob", "bob");

var client = identityManager.AddClient("myapp", "myapp", "somePassword")
    .AddAllowedGrantType("authorization_code")
    .AddAllowedScope("openid")
    .AddAllowedScope("profile")
    .AddAllowedScope("resource");

var web = builder.AddProject<Projects.WebApp>("web")
    .WithReference(client);

builder.Build().Run();
```

I modeled the Client as a child resource (like a database for a container) so that I can pass it to the app and have a custom component configure the Authentication section. It works but I would love to hide the resource from the dashboard as its not really needed.

```csharp
public static IResourceBuilder<IdentityManagerClientResource> AddClient(this IResourceBuilder<IdentityManagerResource> builder, string name, string clientId, string clientSecret)
{
    var client = new IdentityManagerClientResource(name, clientId, clientSecret, builder.Resource);
    
    builder.Resource.Clients.Add(client);
    
    return builder.ApplicationBuilder.AddResource(client).WithInitialState(new()
    {
        ResourceType = "IdentityManagerClient",
        Properties = [],
        State = KnownResourceStates.Hidden
    });
}
```

I tried with `WithInitialState` but the resource still shows on the dashboard

![image](https://github.com/user-attachments/assets/2f2408f2-9ed9-4ec2-ab94-9308d7446b48)

**Comment by davidfowl** [comment]
What does `IdentityManagerClientResource` derive from? Marking it hidden should hide it from the dashboard.

**Reply by Kralizek** [reply]
Here are the two resources


```
public class IdentityManagerResource(string name) : ContainerResource(name), IResourceWithServiceDiscovery, IResourceWithEnvironment, IResourceWithEndpoints {}
public class IdentityManagerClientResource(string name) : Resource(ThrowIfNull(name)), IResourceWithParent<IdentityManagerResource> {}
```

**Reply by Kralizek** [reply]
@davidfowl any extra info I can provide regarding this problem?

**Comment by jdasilva-olo** [comment]
I didn't dig into the code, but as far as I can tell, parent resources publish state updates to their child resources which causes them to become un-hidden.  I've found two strategies for this:

1. Not to add the resource to the application builder.  For example, I'm creating a CustomConnectionStringResource to enable modifying a connection string and then pass it around as if it were a resource.  I don't need it in the manifest and I don't need to see it in Dashboard, so I think it's going to be fine to treat it as a "pseudo" resource that never gets added.

2. Use a an `IDistributedApplicationLifecycleHook` to keep it hidden.  This seems to work, so does updating the state once in `AfterResourcesCreatedAsync` though that would show the child resource for a bit if the app takes long to start.

```
public class KeepHiddenLifecycleHook(ResourceNotificationService notificationService, Resource resource) : IDistributedApplicationLifecycleHook
{
    public Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken = new CancellationToken())
    {
        _ = Task.Run(async () =>
        {
            await foreach (var resourceEvent in notificationService.WatchAsync(cancellationToken))
            {
                if (resourceEvent.Resource == resource && resourceEvent.Snapshot.State?.Text!=KnownResourceStates.Hidden)
                {
                    await notificationService.PublishUpdateAsync(resource, snapshot => snapshot with
                    {
                        State = new(KnownResourceStates.Hidden, snapshot.State?.Style)
                    });
                }
            }
        }, cancellationToken);

        return Task.CompletedTask;
    }
}
```

**Reply by Kralizek** [reply]
Interesting. Thanks! I'll try not adding it to the builder and see how it goes.

**Reply by akovanev** [reply]
It might make sense to modify the hook to allow multiple resources 
```csharp
internal class KeepHiddenLifecycleHook(ResourceNotificationService notificationService, IResource[] resources)
{
    // ...
    if (resources.Contains(resourceEvent.Resource) && resourceEvent.Snapshot.State?.Text != KnownResourceStates.Hidden)
   // ...
}
```

And then build an extension
```csharp
public static IServiceCollection AddToHiddenResources(this IServiceCollection services, IResource[] resources)
{
    services.AddSingleton<ResourceNotificationService>();
    services.AddSingleton<IDistributedApplicationLifecycleHook>(
        sp =>
        {
            var notificationService = sp.GetRequiredService<ResourceNotificationService>();
            return new KeepHiddenLifecycleHook(notificationService, resources);
        });
    
    return services;
}
```

**Reply by davidfowl** [reply]
We have a new first class property for hidden resources coming in 9.3+

**Reply by Kralizek** [reply]
You mean an annotation? 😜

**Reply by davidfowl** [reply]
Actually we don't have an annotation @mitchdenny @adamint , lets add one 😄

**Reply by cchen-motion** [reply]
@davidfowl I can't find any info on the followup; do you have a pointer to the first class property that was added for this?

**Reply by adamint** [reply]
> @davidfowl I can't find any info on the followup; do you have a pointer to the first class property that was added for this?

https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.applicationmodel.customresourcesnapshot.ishidden?view=dotnet-aspire-9.0
